### PR TITLE
Escape special characters in section name of editor config

### PR DESCRIPTION
**Pull request by captainsafia**
Addresses part of https://github.com/dotnet/roslyn/issues/52469.

### CODE REVIEW

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 1]
```diff
@@ -101,6 +101,35 @@ public override bool Execute()
...
+
+        private static string EncodeString(string p)
+        {
+            StringBuilder builder = new StringBuilder();
+            for (var i = 0; i < p.Length; i++)
+            {
+                builder.Append(p[i] switch
+                {
+                    '*' => "\\*",
+                    '?' => "\\?",
```
Comment (Youssef1313) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:119]: In Windows, these aren't valid in file names. If it's the same in Linux and macOS, then this won't be reachable.

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 2]
```diff
@@ -101,6 +101,35 @@ public override bool Execute()
...
+                    '{' => "\\{",
+                    ',' => "\\,",
+                    ';' => "\\;",
+                    '}' => "\\}",
+                    '[' => "\\[",
+                    ']' => "\\]",
+                    '#' => "\\#",
+                    var @default => @default
```
Comment (Youssef1313) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:127]:
```suggestion
                    _ => p[i]
```

**Comment by Youssef1313** [comment]
> I added the ConfigWithEscapedValues in AnalyzerConfigTests to validate that the SectionNameLexer correctly processes the escaped values but it doesn't seem like the Parse helper in the AnalyzerConfigTests actually calls this codepath? It looks like it stops processing the section name here.

I believe this code path is covered by when you call `TryCreateSectionNameMatcher`, which should be already covered in tests. For example this test:

https://github.com/dotnet/roslyn/blob/cf42e6bdd55b17ca6eaa3fd759cc25ac77f3790d/src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerConfigTests.cs#L401-L411

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:119]: They look valid on Linux. Ignore the previous comment.

File: src/Compilers/Core/MSBuildTaskTests/GenerateMSBuildEditorConfigTests.cs [hunk 3]
```diff
@@ -97,6 +97,34 @@ public void MultipleItemMetaDataCreatesSections()
 ", result);
         }
 
+        [Fact]
```
Comment (Youssef1313) [inline src/Compilers/Core/MSBuildTaskTests/GenerateMSBuildEditorConfigTests.cs:100]: Can you add a WorkItem attribute for this?

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 4]
```diff
@@ -101,6 +101,35 @@ public override bool Execute()
...
+        // Filenames with special characters like '#' and'{' get written
+        // into the section names in the resulting .editorconfig file. Later,
+        // when the file is parsed in configuration options these special
+        // characters are interpretted as invalid values and ignored by the
+        // processor. We encode the special characters in these strings
+        // before writing them here.
+
+        private static string EncodeString(string p)
+        {
+            StringBuilder builder = new StringBuilder();
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:113]: can you use a PooledStringBuilder here?  thanks!

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 5]
```diff
@@ -101,6 +101,35 @@ public override bool Execute()
...
+            for (var i = 0; i < p.Length; i++)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:114]: `i` is only used for indexing, so thsi can just be: `foreach (var c in p) ... c switch ...`

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 6]
```diff
@@ -101,6 +101,35 @@ public override bool Execute()
...
+                {
+                    '*' => "\\*",
+                    '?' => "\\?",
+                    '{' => "\\{",
+                    ',' => "\\,",
+                    ';' => "\\;",
+                    '}' => "\\}",
+                    '[' => "\\[",
+                    ']' => "\\]",
+                    '#' => "\\#",
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:126]: a simpler approach here would just be:

```c#
if (c is `'*' or '?' or ... )
    builder.Append("\\");

builder.Append(c);
```

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:113]: @CyrusNajmabadi Why not even avoid creating StringBuilder at all?
The caller already has a StringBuilder that we can reuse.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:113]: that would work for me as well.

File: src/Compilers/Core/MSBuildTaskTests/GenerateMSBuildEditorConfigTests.cs (hunk 3 above)
Comment (captainsafia) [reply src/Compilers/Core/MSBuildTaskTests/GenerateMSBuildEditorConfigTests.cs:100]: Done.

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs (hunk 4 above)
Comment (captainsafia) [reply src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs]: Passed in the `StringBuilder` from the caller.

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 7]
```diff
@@ -101,6 +101,25 @@ public override bool Execute()
             return true;
         }
 
+        // Filenames with special characters like '#' and'{' get written
+        // into the section names in the resulting .editorconfig file. Later,
+        // when the file is parsed in configuration options these special
+        // characters are interpretted as invalid values and ignored by the
+        // processor. We encode the special characters in these strings
+        // before writing them here.
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:109]: just use a normal doc comment.

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 8]
```diff
@@ -101,6 +101,25 @@ public override bool Execute()
...
+
+        private static void EncodeString(StringBuilder builder, string p)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:111]:
```suggestion
        private static void EncodeString(StringBuilder builder, string value)
```

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 9]
```diff
@@ -101,6 +101,25 @@ public override bool Execute()
...
+        {
+            foreach (var c in p)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:113]:
```suggestion
            foreach (var c in value)
```

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 10]
```diff
@@ -101,6 +101,27 @@ public override bool Execute()
...
+
+        private static void EncodeString(StringBuilder builder, string value)
+        {
+            foreach (var c in value)
+            {
+                if (c is '*' or '?' or '{' or ',' or ';' or '}' or '[' or ']' or '#')
+                {
+                    builder.Append("\\");
+                }
+                builder.Append(c);
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:121]: For this scenario, I tried to stick to values that were defined as special characters per the editor config specification (http://docs.editorconfig.org/en/master/editorconfig-format.html).

**Review by chsienki — APPROVED** [review]
LGTM, Thanks!

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs [hunk 11]
```diff
@@ -101,6 +101,27 @@ public override bool Execute()
...
+        /// characters are interpretted as invalid values and ignored by the
+        /// processor. We encode the special characters in these strings
+        /// before writing them here.
+        /// </remarks>
+
+        private static void EncodeString(StringBuilder builder, string value)
+        {
+            foreach (var c in value)
+            {
+                if (c is '*' or '?' or '{' or ',' or ';' or '}' or '[' or ']' or '#')
```
Comment (cston) [inline src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:117]: >c is [](http://example.com/codeflow?start=20&length=4)

Do we need to include `!`?

Comment (captainsafia) [reply src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs:117]: Yep, it is a wildcard character. I'll go ahead and add it.

File: src/Compilers/Core/MSBuildTaskTests/GenerateMSBuildEditorConfigTests.cs [hunk 12]
```diff
@@ -97,6 +97,35 @@ public void MultipleItemMetaDataCreatesSections()
...
+        [WorkItem(52469, "https://github.com/dotnet/roslyn/issues/52469")]
+        public void MultipleSpecialCharacterItemMetaDataCreatesSections()
```
Comment (cston) [inline src/Compilers/Core/MSBuildTaskTests/GenerateMSBuildEditorConfigTests.cs:102]: >MultipleSpecialCharacterItemMetaDataCreatesSections [](http://example.com/codeflow?start=20&length=51)

Please consider testing all the special characters from `EncodeString()` if not already tested.

Comment (captainsafia) [reply src/Compilers/Core/MSBuildTaskTests/GenerateMSBuildEditorConfigTests.cs:102]: I've updated the tests to validate the full set of special chars.

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs (hunk 11 above)
Comment (Youssef1313) [reply src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs]: @cston @captainsafia What is the special meaning of `!`? It's treated currently as a simple character:

https://github.com/dotnet/roslyn/blob/481529aa563e63b059b50f826cd99f523d680f57/src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.SectionNameMatching.cs#L443

**Pull request merged**